JuliaSyntax/src/porcelain/syntax_graph.jl

```diff
@@ -857,3 +875,488 @@ function _map_green_to_ast(parent::Kind, cs::SyntaxList; eq_to_kw=false)
+        # (doc str obj) => (macrocall Core.@doc lno str obj)
+        ret_k = K"macrocall"
+        pushfirst!(cs, valleaf(source_location(LineNumberNode, st)))
+        pushfirst!(cs, valleaf(GlobalRef(Core, Symbol("@doc"))))
+    elseif k === K"dotcall" || k === K"call" && n_cs > 0
+        if is_infix_op_call(st) || is_postfix_op_call(st)
+            cs[2], cs[1] = cs[1], cs[2]
+        end
+        if is_postfix_op_call(st) && kind(cs[1]) == K"Identifier" &&
+            cs[1].name_val === "'"
+            popfirst!(cs)
+            ret_k = K"'"
+        end
+        do_ex = kind(cs[end]) === K"do" ? pop!(cs) : nothing
+        _reorder_parameters!(cs, 2)
+        if k === K"dotcall"
+            if is_prefix_call(st)
+                # (dotcall f args...) => (. f (tuple args...))
+                ret_cs = _map_green_to_est(st, cs)
+                tuple = newnode(graph, st, K"tuple", ret_cs[2:end])
+                return newnode(graph, st, K".", tree_ids(ret_cs[1], tuple))
+            else
+                # (dotcall + args...) => (call .+ args...)
+                ret_k = K"call"
+                if kind(cs[1]) === K"Identifier"
+                    cs[1] = symleaf('.' * cs[1].name_val)
+                end
+            end
+        end
+        !isnothing(do_ex) && return _make_do_expression(st, cs, do_ex)
+    elseif k === K"."
+        if n_cs === 2
+            # (. lhs rhs) => (. lhs (inert rhs))
+            lhs = _green_to_est(st, 1, cs[1])
+            rhs = _green_to_est(st, 2, cs[2])
+            inert_rhs = kind(rhs) in KSet"quote inert" ? rhs :
+                newnode(graph, cs[2], K"inert", tree_ids(rhs))
+            return mknode(st, tree_ids(lhs, inert_rhs))
+        elseif n_cs === 1
+            # (. x) => (. x) or .x
+            # TODO: This is the one place where K"parens" change the result,
+            # meaning that either Expr is doing something wrong or SyntaxNode is
+            # deleting semantics.
+            paren_st = filter(should_include_node, children(parent))[1]
+            coalesce_dot = !(kind(paren_st) === K"parens") && parent_i === 1 &&
+                kind(parent) in KSet"call dotcall curly quote"
+
+            if (coalesce_dot || is_syntactic_operator(kind(cs[1])) ||
+                kind(parent) === K"comparison" && iseven(parent_i))
+                return symleaf('.' * cs[1].name_val)
+            end
+        end
+    elseif k === K"ref" || k === K"curly"
+        _reorder_parameters!(cs, 2)
+    elseif k === K"for" && n_cs === 2
+        # (for (iteration iter1) body) => (for iter1 body)
+        iters = preprocessed_green_children(cs[1])
+        if length(iters) === 1
+            cs[1] = iters[1]
+        end
+    elseif k === K"iteration"
+        # (for (iteration iter1 iters...) body) => (for (block iter1 iters...) body)
+        @assert kind(parent) === K"for" && parent_i === 1
+        ret_k = K"block"
+    elseif k === K"tuple" || k === K"vect" || k === K"braces"
```

Comment:
very unimportant: if you split `K"tuple"` you can lose one extra line of scope here